### PR TITLE
[test-framework] Improve Kubernetes and Mixer support.

### PR DESCRIPTION
**Pull request by ozevren**
+ Changed istio-system command-line flag and calculation behavior, so that the default system namespace is "istio-system" when nothing is specified. As we make more progress on being able to deploy dynamically to non-"istio-system" namespaces, we can change the behavior.
+ Added generic DumpState support to environment interface, to allow dumping of state of later investigation. This should be particularly useful in CI environments.
+ Sprinkled calls to DumpState everywhere. Ideally we should have a centralized way to do this. Unfortunately, I haven't figured out a clean way
to do this holistically yet.
+ Improved Pod Data dump to include both the pod logs, as well as the current snapshot of the Pod resource in Yaml format. Both get dumped to the work directory, in case of failures.
+ Fine-tuned and normalized retry timing characteristics for all Kubernetes operations.
+ Added logic to extract attribute manifest directly from the Helm charts. The attribute manifest needs to be injected during local environment runs, but is not needed in the Kubernetes environment. This approach allows us to have identical configuration for Mixer that can be used in both Local and Kubernetes environments.
+ Added a retry logic to Mixer bypass adapter's connectivity logic to help issues with initial connectivity after config push.
+ Introduced framework/repository as a place for common constants for Istio source repository.
+ Removed the logic that deletes collected reports from the PolicyBackend fake, to reduce state management that needs to be done in the TestFramework.
+ Updated PolicyBackend's APIs to do a subset query of reports, instead of exact match. This is needed, as unrelated telemetry is also collected (i.e. Mixer's own telemetry) during testing.
+ Added retry logic in a few places in the Kubernetes code to deal with eventual consistency.
+ Added template expansion support to Mixer instances.
+ Fixed the issue with PolicyBackend deployment in CircleCI environment (ImagePullPolicy Always => IfNotPresent)
+ Fixed basic Mixer tests to work in both environments.

### CODE REVIEW

**Comment by smg223** [comment]
Can u tell me how to use a github repository to detect imsi-catchers?

File: mixer/adapter/bypass/bypass.go [hunk 1]
```diff
@@ -159,12 +160,17 @@ func (b *builder) Build(context context.Context, env adapter.Env) (adapter.Handl
...
 func (b *builder) ensureConn() error {
+	const timeOut = time.Minute
+	const delay = time.Millisecond * 10
+
 	if b.conn == nil {
-		conn, err := grpc.Dial(b.params.BackendAddress, grpc.WithInsecure())
-		if err != nil {
-			return err
+		var err error
+		for start := time.Now(); time.Now().Before(start.Add(timeOut)); {
```
Comment (nmittler) [inline mixer/adapter/bypass/bypass.go]: We have a few retry libraries lying around ... maybe use one of those?

Also, there's a pretty lightweight lib for this sort of thing: https://github.com/cenkalti/backoff/blob/master/retry.go

File: pkg/test/framework/components/mixer/attrmanifest.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	"errors"
+	"strings"
+
+	"istio.io/istio/pkg/test/framework/repository"
+	"istio.io/istio/pkg/test/framework/scopes"
+	"istio.io/istio/pkg/test/helm"
+)
+
+// Extract attribute manifest from Helm charts.
+func extractAttributeManifestFromHelm() (string, error) {
```
Comment (nmittler) [inline pkg/test/framework/components/mixer/attrmanifest.go]: Is this mixer-specific? If not, maybe it should be in a more common place? Maybe something like `/pkg/test/kube/helm/...`?

File: pkg/test/framework/components/mixer/attrmanifest.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		"attributemanifest",
+		"istio-system",
+		repository.IstioChartDir(),
+		"", nil)
+	if err != nil {
+		return "", err
+	}
+
+	// Split the template into chunks
+	parts := strings.Split(s, "\n---\n")
```
Comment (nmittler) [inline pkg/test/framework/components/mixer/attrmanifest.go]: Is this notation specific to k8s, or is it a general yaml separator? We have a lot of this code lying around ... would be good to use a utility for this.

File: pkg/test/framework/components/policybackend/policybackend.go [hunk 4]
```diff
@@ -84,8 +85,8 @@ spec:
     spec:
       containers:
       - name: app
-        image: {{.Hub}}/test_policybackend:{{.Tag}}
-        imagePullPolicy: {{.ImagePullPolicy}}
+        image: "{{.Hub}}/test_policybackend:{{.Tag}}"
+        imagePullPolicy: IfNotPresent
```
Comment (nmittler) [inline pkg/test/framework/components/policybackend/policybackend.go]: Why hard-coding? For the test framework, I would think we would want `Always` by defaultt to ensure we're always using the latest image.

File: pkg/test/framework/components/policybackend/policybackend.go [hunk 5]
```diff
@@ -44,15 +47,13 @@ var (
 	// KubeComponent is a component for the Kubernetes environment.
 	KubeComponent = &kubeComponent{}
 
-	scope                                   = log.RegisterScope("policybackend", "Policy backend test component", 0)
-	_     environment.DeployedPolicyBackend = &policyBackend{}
-	_     io.Closer                         = &policyBackend{}
+	_ environment.DeployedPolicyBackend = &policyBackend{}
```
Comment (nmittler) [inline pkg/test/framework/components/policybackend/policybackend.go]: why do we need these? Is this just to ensure that casting works?

File: pkg/test/framework/environment/environment.go [hunk 6]
```diff
@@ -110,6 +110,9 @@ type (
 	Implementation interface {
 		// EnvironmentID is the unique ID of the implemented environment.
 		EnvironmentID() settings.EnvironmentID
+
+		// Evaluate the given template with environment specific template variables.
+		Evaluate(tmpl string) (string, error)
```
Comment (nmittler) [inline pkg/test/framework/environment/environment.go]: Just a note: I suspect we'll have to think a bit about this ... we need a different approach here going forward. The tests shouldn't have environment-specific things in them.

File: pkg/test/framework/repository/paths.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package repository have utility methods related to accessing a local istio code repository. The repository
+// path is deduced from ${GOPATH} environment variable.
+package repository
```
Comment (nmittler) [inline pkg/test/framework/repository/paths.go]: This file seems somewhat redundant with https://github.com/istio/istio/blob/master/pkg/test/env/istio.go.

I suggest adding the helm-specific variables there.

File: pkg/test/kube/port_forwarder.go [hunk 8]
```diff
@@ -60,6 +62,7 @@ type PodSelectOptions struct {
 }
 
 func (f *defaultPortForwarder) Start() error {
+	scopes.Framework.Debugf("starting port forwarder")
```
Comment (nmittler) [inline pkg/test/kube/port_forwarder.go]: how is this helpful?

File: pkg/test/framework/components/mixer/attrmanifest.go (hunk 3 above)
Comment (ozevren) [reply pkg/test/framework/components/mixer/attrmanifest.go]: Yaml specific: http://yaml.org/spec/1.0/ (section 2.2).
Moved to the top-level "test" library.

**Comment by geeknoid** [comment]
/lgtm
/approve

Mixer parts look good.

**Pull request merged**